File: app/lib/e2e/trustlog/codec.dart
import 'dart:typed_data';

import '../bytes.dart' show compareBytes, putUint32;
import '../ed25519.dart' show ed25519Verify;
import '../symmetric_state.dart' show blake2s;
import 'entry.dart';

const int _maxField = 1 << 20;
const int _maxSigners = 1 << 12;
const int _maxDisablements = 1 << 12;
const int _maxCoSigns = 1 << 12;
const int _maxReplaces = 1 << 12;
const int _maxEntries = 1 << 16;

/// Returns e's co-signs sorted by signer, then by sig — the canonical order Go
/// trustlog.canonicalCoSigns produces. Co-signs are covered by hashEntry but
/// their order is committed by nothing, so a gateway could otherwise reorder them
/// to make the client compute a different Tip than the node. Returns a sorted
/// copy; the input is not mutated.
List<CoSign> _canonicalCoSigns(List<CoSign> cs) {
  final out = List<CoSign>.of(cs);
  out.sort((a, b) {
    final c = compareBytes(a.signer, b.signer);
    if (c != 0) return c;
    return compareBytes(a.sig, b.sig);
  });
  return out;
}

/// The chain hash of a full (signed) entry. Covers sigBytes ‖ putField(sig).
/// For KindRevokeSigner, replaces is already in sigBytes; CoSigns are appended
/// after Sig in canonical order so the chain commits to the full co-signed
/// payload regardless of co-sign order (mirrors Go hashEntry).
Uint8List hashEntry(Entry e) {
  final b = BytesBuilder();
  b.add(sigBytes(e));
  putField(b, e.sig);
  if (e.kind == Kind.revokeSigner) {
    final cs = _canonicalCoSigns(e.coSigns);
    putUint32(b, cs.length);
    for (final c in cs) {
      putField(b, c.signer);
      putField(b, c.sig);
    }
  }
  return blake2s(b.toBytes());
}

/// Returns the number of distinct valid co-signs from trusted (non-revoked) signers,
/// and whether that count exceeds the number of revoked signers. By default the
/// revoked signers (e.signers) are excluded; when [allowRevoked] is true they may
/// co-sign (KindRevokeSigner with replacements — voluntary rotation).
/// Mirrors Go validCoSigns exactly.
Future<(int, bool)> validCoSigns(Entry e, bool Function(Uint8List) trusted, {bool allowRevoked = false}) async {
  final sb = sigBytes(e);
  // revoked set: signers listed in e.signers (the revoked pubkeys)
  final revoked = <String>{};
  if (!allowRevoked) {
    for (final r in e.signers) {
      revoked.add(String.fromCharCodes(r));
    }
  }
  final seen = <String>{};
  var n = 0;
  for (final cs in e.coSigns) {
    final ks = String.fromCharCodes(cs.signer);
    if (seen.contains(ks) || revoked.contains(ks) || !trusted(cs.signer)) continue;
    // Total-function verify: a malformed (wrong-length) co-sign is skipped, not
    // thrown — matching Go validCoSigns, so a junk co-sign can't reject the chain.
    if (!await ed25519Verify(cs.signer, sb, cs.sig)) continue;
    seen.add(ks);
    n++;
  }
  return (n, n > e.signers.length);
}

class _Reader {
  _Reader(this._b) : _view = ByteData.sublistView(_b);
  final Uint8List _b;
  final ByteData _view;
  int _off = 0;
  int get remaining => _b.length - _off;

  int readU8() {
    if (remaining < 1) throw const FormatException('trustlog: truncated');
    return _b[_off++];
  }

  int readU32() {
    if (remaining < 4) throw const FormatException('trustlog: truncated');
    final v = _view.getUint32(_off, Endian.big);
    _off += 4;
    return v;
  }

  int readCount(int cap) {
    final n = readU32();
    if (n > cap) throw const FormatException('trustlog: count exceeds cap');
    return n;
  }

  Uint8List? readField() {
    final n = readU32();
    if (n > _maxField) throw const FormatException('trustlog: field exceeds cap');
    if (n == 0) return null;
    if (remaining < n) throw const FormatException('trustlog: truncated field');
    final f = Uint8List.fromList(Uint8List.sublistView(_b, _off, _off + n));
    _off += n;
    return f;
  }
}

Entry _readEntry(_Reader r) {
  final kind = Kind.fromValue(r.readU8());
  if (kind == null) throw const FormatException('trustlog: unknown kind');
  final prev = r.readField();
  final sc = r.readCount(_maxSigners);
  final signers = [for (var i = 0; i < sc; i++) r.readField() ?? Uint8List(0)];
  final dc = r.readCount(_maxDisablements);
  final disablements = [for (var i = 0; i < dc; i++) r.readField() ?? Uint8List(0)];
  final key = r.readField();
  final signer = r.readField();
  // KindRevokeSigner: Replaces count+fields come after Signer (part of sigBytes) and before Sig.
  List<Uint8List> replaces = const [];
  if (kind == Kind.revokeSigner) {
    final rc = r.readCount(_maxReplaces);
    replaces = [for (var i = 0; i < rc; i++) r.readField() ?? Uint8List(0)];
  }
  final sig = r.readField();
  List<CoSign> coSigns = const [];
  if (kind == Kind.revokeSigner) {
    final csc = r.readCount(_maxCoSigns);
    coSigns = [
      for (var i = 0; i < csc; i++)
        CoSign(signer: r.readField() ?? Uint8List(0), sig: r.readField() ?? Uint8List(0))
    ];
  }
  return Entry(
      kind: kind, prev: prev, signers: signers, disablements: disablements,
      key: key, signer: signer, sig: sig, coSigns: coSigns, replaces: replaces);
}

Entry unmarshalEntry(Uint8List b) {
  final r = _Reader(b);
  final e = _readEntry(r);
  if (r.remaining != 0) throw const FormatException('trustlog: trailing bytes after entry');
  return e;
}

List<Entry> unmarshalChain(Uint8List b) {
  final r = _Reader(b);
  final cnt = r.readCount(_maxEntries);
  final entries = <Entry>[];
  for (var i = 0; i < cnt; i++) {
    final raw = r.readField();
    if (raw == null) throw const FormatException('trustlog: empty entry');
    entries.add(unmarshalEntry(raw));
  }
  if (r.remaining != 0) throw const FormatException('trustlog: trailing bytes after chain');
  return entries;
}

File: internal/trustlog/codec.go
package trustlog

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// Decode caps: chains arrive over the untrusted gateway relay, so every
// attacker-controllable length/count is bounded before allocation.
const (
	maxField        = 1 << 20 // 1 MiB per field (keys/sigs/hashes are tiny; generous)
	maxSigners      = 1 << 12 // signers in a genesis
	maxDisablements = 1 << 12 // disablement commitments in a genesis
	maxEntries      = 1 << 16 // entries in a chain
	maxCoSigns      = 1 << 12 // co-signs in a KindRevokeSigner entry
	maxReplaces     = 1 << 12 // replacement signers in a KindRevokeSigner entry
)

// MarshalEntry encodes an entry to its canonical wire form. It is identical to the
// bytes hashEntry covers, so decode→hash reproduces the original chain hash.
// For KindRevokeSigner, sigBytes already covers Replaces, and CoSigns are
// appended after Sig (mirroring hashEntry).
func MarshalEntry(e Entry) []byte {
	var buf bytes.Buffer
	buf.Write(sigBytes(&e))
	putField(&buf, e.Sig)
	if e.Kind == KindRevokeSigner {
		var cnt [4]byte
		binary.BigEndian.PutUint32(cnt[:], uint32(len(e.CoSigns)))
		buf.Write(cnt[:])
		for _, cs := range e.CoSigns {
			putField(&buf, cs.Signer)
			putField(&buf, cs.Sig)
		}
	}
	return buf.Bytes()
}

func getField(r *bytes.Reader) ([]byte, error) {
	var n uint32
	if err := binary.Read(r, binary.BigEndian, &n); err != nil {
		return nil, err
	}
	if n > maxField {
		return nil, fmt.Errorf("trustlog: field length %d exceeds cap", n)
	}
	if n == 0 {
		return nil, nil
	}
	b := make([]byte, n)
	if _, err := io.ReadFull(r, b); err != nil {
		return nil, err
	}
	return b, nil
}

func getCount(r *bytes.Reader, cap uint32) (uint32, error) {
	var n uint32
	if err := binary.Read(r, binary.BigEndian, &n); err != nil {
		return 0, err
	}
	if n > cap {
		return 0, fmt.Errorf("trustlog: count %d exceeds cap", n)
	}
	return n, nil
}

// UnmarshalEntry decodes a canonical entry. It rejects truncation, oversized
// prefixes, and trailing bytes.
func UnmarshalEntry(b []byte) (Entry, error) {
	r := bytes.NewReader(b)
	e, err := readEntry(r)
	if err != nil {
		return Entry{}, err
	}
	if r.Len() != 0 {
		return Entry{}, errors.New("trustlog: trailing bytes after entry")
	}
	return e, nil
}

func readEntry(r *bytes.Reader) (Entry, error) {
	kind, err := r.ReadByte()
	if err != nil {
		return Entry{}, err
	}
	var e Entry
	e.Kind = Kind(kind)
	if e.Prev, err = getField(r); err != nil {
		return Entry{}, err
	}
	cnt, err := getCount(r, maxSigners)
	if err != nil {
		return Entry{}, err
	}
	if cnt > 0 {
		e.Signers = make([][]byte, cnt)
		for i := range e.Signers {
			if e.Signers[i], err = getField(r); err != nil {
				return Entry{}, err
			}
		}
	}
	dcnt, err := getCount(r, maxDisablements)
	if err != nil {
		return Entry{}, err
	}
	if dcnt > 0 {
		e.Disablements = make([][]byte, dcnt)
		for i := range e.Disablements {
			if e.Disablements[i], err = getField(r); err != nil {
				return Entry{}, err
			}
		}
	}
	if e.Key, err = getField(r); err != nil {
		return Entry{}, err
	}
	if e.Signer, err = getField(r); err != nil {
		return Entry{}, err
	}
	if e.Kind == KindRevokeSigner {
		rcnt, err := getCount(r, maxReplaces)
		if err != nil {
			return Entry{}, err
		}
		if rcnt > 0 {
			e.Replaces = make([][]byte, rcnt)
			for i := range e.Replaces {
				if e.Replaces[i], err = getField(r); err != nil {
					return Entry{}, err
				}
			}
		}
	}
	if e.Sig, err = getField(r); err != nil {
		return Entry{}, err
	}
	if e.Kind == KindRevokeSigner {
		cscnt, err := getCount(r, maxCoSigns)
		if err != nil {
			return Entry{}, err
		}
		if cscnt > 0 {
			e.CoSigns = make([]CoSign, cscnt)
			for i := range e.CoSigns {
				if e.CoSigns[i].Signer, err = getField(r); err != nil {
					return Entry{}, err
				}
				if e.CoSigns[i].Sig, err = getField(r); err != nil {
					return Entry{}, err
				}
			}
		}
	}
	return e, nil
}

// MarshalChain encodes a sequence of entries: a count then each entry length-prefixed.
func MarshalChain(entries []Entry) []byte {
	var buf bytes.Buffer
	var cnt [4]byte
	binary.BigEndian.PutUint32(cnt[:], uint32(len(entries)))
	buf.Write(cnt[:])
	for i := range entries {
		putField(&buf, MarshalEntry(entries[i]))
	}
	return buf.Bytes()
}

// UnmarshalChain decodes a chain produced by MarshalChain, rejecting garbage,
// truncation, oversized counts/fields, and trailing bytes.
func UnmarshalChain(b []byte) ([]Entry, error) {
	r := bytes.NewReader(b)
	cnt, err := getCount(r, maxEntries)
	if err != nil {
		return nil, err
	}
	var entries []Entry
	for i := uint32(0); i < cnt; i++ {
		raw, err := getField(r)
		if err != nil {
			return nil, err
		}
		e, err := UnmarshalEntry(raw)
		if err != nil {
			return nil, fmt.Errorf("trustlog: entry %d: %w", i, err)
		}
		entries = append(entries, e)
	}
	if r.Len() != 0 {
		return nil, errors.New("trustlog: trailing bytes after chain")
	}
	return entries, nil
}
